base/operators.jl

```diff
@@ -818,19 +818,83 @@ function setindex_shape_check{T}(X::AbstractArray{T,2}, i::Integer, j::Integer)
 end
 setindex_shape_check(X, I...) = nothing # Non-arrays broadcast to all idxs
 
-# convert to a supported index type (Array, Colon, or Int)
-to_index(i::Int) = i
+# convert to a supported index type (array or Int)
+"""
+    to_index(A, i)
+
+Convert index `i` for use in indexing into array `A`. Custom array types
+should specialize `to_index(::CustomArray, i)` to provide special indexing
+behaviors. Note that some index types (like `Colon`) require more context in
+order to transform them into an array of indices; those get converted in the
+more complicated `to_indices` function. By default, this simply calls the
+generic `to_index(i)`.
+"""
+to_index(A, i) = to_index(i)
+
+"""
+    to_index(i)
+
+Convert index `i` to an `Int` or array of indices to be used as an index for all
+arrays.  Custom index types should specialize `to_index(::CustomIndex)` to
+provide special indexing behaviors.
+"""
 to_index(i::Integer) = convert(Int,i)::Int
-to_index(c::Colon) = c
-to_index(I::AbstractArray{Bool}) = find(I)
-to_index(A::AbstractArray) = A
-to_index{T<:AbstractArray}(A::AbstractArray{T}) = throw(ArgumentError("invalid index: $A"))
-to_index(A::AbstractArray{Colon}) = throw(ArgumentError("invalid index: $A"))
+to_index(I::AbstractArray{Bool}) = LogicalIndex(I)
+to_index(I::AbstractArray) = I
+to_index{T<:Union{AbstractArray, Colon}}(I::AbstractArray{T}) = throw(ArgumentError("invalid index: $I"))
+to_index(::Colon) = throw(ArgumentError("colons must be converted by to_indices(...)"))
 to_index(i) = throw(ArgumentError("invalid index: $i"))
 
-to_indexes() = ()
-to_indexes(i1) = (to_index(i1),)
-to_indexes(i1, I...) = (@_inline_meta; (to_index(i1), to_indexes(I...)...))
+# The general to_indices is mostly defined in multidimensional.jl, but this
+# definition is required for bootstrap:
+"""
+    to_indices(A, I::Tuple)
+
+Convert a tuple of indices `I` for use in indexing into array `A`, where the
+returned indices are either `Int`s or `AbstractArray`s of `Int` or
+`CartesianIndex`. This function accepts any index types and must return a tuple
+containing `Int`s and `AbstractArray`s.
+
+For simple index types, this simply ends up calling `to_index(A, i)` for each
+index `i`. More complicated index types, however, require more context about
+the dimension into which they index. To support those cases, `to_indices(A, I)`
+calls `to_indices(A, indices(A), I)`, which then recursively walks through both
+the given tuple of indices and the dimensional indices of `A` in tandem.
+"""
+to_indices(A, I::Tuple) = (@_inline_meta; to_indices(A, indices(A), I))
```

Comment:
No, but the advantage is that it allows for a simpler `to_index(A, i)` method for array types to override.